Add session expiry

kokonutwh1skey opened the pull request:
1. Add session expiry
2. Reorganize files
3. Fix typos in `README.md`

AlexeyAkhunov commented:
thank you! I had a quick look and there are couple of issues you may want to look into:
1. There are some stray dependencies in `go.sum` which you may want to clean up using `go mod tidy`
2. The expiration is currently only checked when the the session is being accessed, and this is not a good time to expire the session, because it means someone needs it. This also does not clean up the sessions that have not been accessed for a long time.

I suggest that you try to organise sessions in a priority queue, prioritised by the time since the last access (you can use `container/heap` for that), and introduce a configurable limit of how many active sessions there may be at any given time. That way, oldest sessions will be removed when room for a new session is required.

kokonutwh1skey commented:
> thank you! I had a quick look and there are couple of issues you may want to look into:
> 
> 1. There are some stray dependencies in `go.sum` which you may want to clean up using `go mod tidy`
> 2. The expiration is currently only checked when the the session is being accessed, and this is not a good time to expire the session, because it means someone needs it. This also does not clean up the sessions that have not been accessed for a long time.
> 
> I suggest that you try to organise sessions in a priority queue, prioritised by the time since the last access (you can use `container/heap` for that), and introduce a configurable limit of how many active sessions there may be at any given time. That way, oldest sessions will be removed when room for a new session is required.

I think LRU is a good candidate to implement it.

AlexeyAkhunov commented:
> I think LRU is a good candidate to implement it.

Yes, definitely LRU cache from a library can also be used

kokonutwh1skey commented:
Updated PR with LRU

AlexeyAkhunov commented:
Should UiSessions also be turned into LRU cache? Is the lock for the nodeSessionMap still required? Default value of `max.sessions` of 50 is way too low, I would say 5000 :)

AlexeyAkhunov commented:
Can you also email me at `tg@torquem.ch` quoting your GitHub name, I need to check something with you

kokonutwh1skey commented:
> Should UiSessions also be turned into LRU cache? Is the lock for the nodeSessionMap still required? Default value of `max.sessions` of 50 is way too low, I would say 5000 :)

Yeah, I missed the `UiSessions`.

Updated pr:
1. LRU for ui sessions
2. Remove locks for both uiSessions and nodeSessions maps
3. Changed the default value to 5000

AlexeyAkhunov commented:
Thank you, I will take a look shortly. Now that we have linter, could you please also fix lint errors?

kokonutwh1skey commented:
Fixed lint errors

AlexeyAkhunov commented:
Thanks, can you also tidy up `go.sum` with `go mod tidy`?

kokonutwh1skey commented:
> Thanks, can you also tidy up `go.sum` with `go mod tidy`?

I ran `go mod tidy` but nothing changed.
